url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/handlers/FileHandler.java

```diff
@@ -0,0 +1,212 @@
+package stv6.handlers;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.ParseException;
+import java.util.Date;
+
+import stv6.http.HttpServer;
+import stv6.http.request.Request;
+import stv6.http.request.RequestHandler;
+import stv6.http.request.Response;
+
+/**
+ * This handler serves FILES. By that I mean text files
+ *  (For CSS, for example) and image files. It is a very
+ *  simplistic implementation for images; I have a list
+ *  of extensions and if the file extension matches that,
+ *  it is displayed as an image.
+ * 
+ * @author Daniel
+ *
+ */
+public class FileHandler implements RequestHandler {
+	public static final String HANDLED_PAGE = "files";
+	
+	private static String imgExts[] = {"jpg", "jpeg", "png", "gif", "ico"};
+//	private File theFile;
+	
+	@Override
+	public boolean handle(Request request, Response response) {
+		// skip the /files/
+		String path = request.path
+		    .replace('/', File.separatorChar)
+		    .replace('+', ' ')
+		    .replace("%20", " ").substring(7);
+		File theFile = new File(path);
+		if (!theFile.exists()) {		
+			System.err.println("Cannot find: " + path + " (" +
+			        theFile.getAbsolutePath() + ")");
+			return false;
+		}
+		if (!theFile.canRead()) {
+			System.err.println("Cannot read: " + path);
+			return false;
+		}
+		
+		// can we let them use their cache?
+		if (handleCache(request, response, theFile))
+		    return true;
+		
+		String ext = getImageExt(path);
+		if (ext != null) {
+
+			response.setContentType("image/"+ext);
+			byte[] imgdata = null;
+			try {
+				imgdata = getBytesFromFile(theFile);
+			} catch (IOException e) {
+				e.printStackTrace();
+				return false;
+			}
+				
+			if (imgdata == null)
+				return false;
+			
+			response.setBody(imgdata);
+			return true;
+		} else {
+			// it's just a text file; read it
+			try {		
+
+				BufferedReader reader = new BufferedReader(new FileReader(theFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 77, 80. The resource is closed at line 84. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).